src/storage.rs

```diff
@@ -291,64 +356,42 @@ impl MemStorage {
 impl Storage for MemStorage {
     /// Implements the Storage trait.
     fn initial_state(&self) -> Result<RaftState> {
-        let core = self.rl();
-        let mut state = RaftState {
-            hard_state: core.hard_state.clone(),
-            conf_state: core.snapshot.get_metadata().get_conf_state().clone(),
-            pending_conf_state: None,
-            pending_conf_state_start_index: None,
-        };
-        if core.snapshot.get_metadata().has_pending_membership_change() {
-            state.pending_conf_state = core
-                .snapshot
-                .get_metadata()
-                .get_pending_membership_change()
-                .clone()
-                .into();
-            state.pending_conf_state_start_index = core
-                .snapshot
-                .get_metadata()
-                .get_pending_membership_change_index()
-                .into();
-        }
-        Ok(state)
+        Ok(self.rl().raft_state.clone())
     }
 
     /// Implements the Storage trait.
     fn entries(&self, low: u64, high: u64, max_size: impl Into<Option<u64>>) -> Result<Vec<Entry>> {
         let max_size = max_size.into();
         let core = self.rl();
-        let offset = core.entries[0].get_index();
-        if low <= offset {
+        if low < core.inner_first_index() {
             return Err(Error::Store(StorageError::Compacted));
         }
 
         if high > core.inner_last_index() + 1 {
-            panic!(
```

Comment:
I prefer panicing here as it's unreachable. Reach this line means either storage or raft is not implemented correctly, return `Unavailable` is misleading.

Reply:
Here I think return an Error is better because `entries` is a public API. Users may use the interface in a wrong way. For that maybe an Error is better than panic.

Reply:
How about documenting it clearly in the trait method? Correctness is more important.

Reply:
Done. PTAL thanks.